src/app

```diff
@@ -0,0 +1,131 @@
+    application_name=$(jq -r .application_name < "$1")
+    platform=$(jq -r .platform < "$1")
+    orchestrator=$(jq -r .orchestrator < "$1")
+    version=$(jq -r .version < "$1")
+    app_sub_module="${app_module_dir}/${orchestrator}"
+    if test ! -f "${app_sub_module}"; then
+        echo "ERROR: ${app_sub_module} not found. exiting."
+        return 1
+    fi
+    artifact_name=$(jq -r .artifact_provides.artifact_name < "$2")
+}
+
+decompress_artifact() {
+    local image_dir
+    local image
+    local url_new
+    local url_current
+    local sha_new
+    local sha_current
+    local current_image
+    local new_image
+    local rollback_id="last"
+
+    if test "$temp_dir" = ""; then
+        return 1
+    fi
+
+    $tar_decompress_cmd "$1"/images.tar.gz -C "$temp_dir"
+    $tar_decompress_cmd "$1"/manifests.tar.gz -C "$temp_dir"
+
+    for image_dir in "${temp_dir}/images/"*; do
+        url_new=$(cat "${image_dir}/url-new.txt")
+        url_current=$(cat "${image_dir}/url-current.txt")
+        sha_new=$(cat "${image_dir}/sums-new.txt")
+        sha_current=$(cat "${image_dir}/sums-current.txt")
+        if test "$url_new" != "$url_current"; then
+            current_image="${temp_dir}/current.${sha_current}.img"
+            new_image="${temp_dir}/new.${sha_new}.img"
+            $app_sub_module save "$url_current" "$current_image" "${application_name}"
+            $delta_cmd "$current_image" "${image_dir}/image.img" "${new_image}"
+            mv -v "${new_image}" "${image_dir}/image.img"
+        fi
+    done
+    for image_dir in "${temp_dir}/images/"*; do
+        url_new=$(cat "${image_dir}/url-new.txt")
+        sha_new=$(cat "${image_dir}/sums-new.txt")
+        $app_sub_module load "${url_new}" "${image_dir}/image.img" "${application_name}"
+    done
+    # we should check if the app is healthy and alive, then decide what to do
+    # at the moment we assume it is alive and ok
+    if test -d "${persistent_store}/${application_name}-${rollback_id}"; then
+        rm -Rf "${persistent_store}/${application_name}-${rollback_id}"
+    fi
+    if test -d "${persistent_store}/${application_name}/"; then
+        mv -v "${persistent_store}/${application_name}/" "${persistent_store}/${application_name}-${rollback_id}"
+        $app_sub_module stop "${persistent_store}/${application_name}-${rollback_id}/manifests" "${application_name}"
+    fi
+    mkdir -p "${persistent_store}/${application_name}"
+    mv -v "$temp_dir/manifests" "${persistent_store}/${application_name}/"
+    $app_sub_module rollout "${persistent_store}/${application_name}/manifests" "${application_name}"
+    # now if $app_sub_module healthy then rm -Rf "${persistent_store}/${artifact_name}-${rollback_id}"
+    # and on rollback we just deploy "${persistent_store}/${artifact_name}-${rollback_id}"
+    # earlier ln -s "${persistent_store}/${artifact_name}-${rollback_id}" "${persistent_store}/${artifact_name}-rollback"
+    # and removing all else
```

Comment:
This I didn't fully understand at first glance. I need to come back to it :sweat_smile:

Reply:
let's have a relaxing tea over a call (or a call over relaxing tea).